### DISCUSSION TITLE
[Help] how to extract exe from Zip and then use it for installation

### DISCUSSION THREAD

**Discussion by Xelbayria** [ScoopInstaller/Scoop#5052] [Help and Questions] [answered]
I apologize for the inconvenience. I am stuck at a step where I have no idea how to get it done. The program.EXE is packed into a zip folder... Normally, I would unpack the zip to get EXE file out of the zip folder, then I click it to install the program. 
I have gone through the documentation and a few examples. I wasn't able to find a good example. 

Within the ZIP folder, There is only one file "program.v2.1.2.build201.exe" (the version and build are just an example). I appreciate your help.

**Comment by rashil2000** [comment]
Please use the Discussions tab for help and QnA.

Do you have a download link for the program?

**Reply by Xelbayria** [reply]
Ah i don't know Discussions Tab can be used for such a topic. and here's the [Link to the program](https://www.inputdirector.com/downloads/InputDirector.v2.1.2.zip)

**Comment by niheaven** [comment]
I've made one, and you could ref it.

```json
"url": "https://cdn.apifox.cn/download/Apifox-windows-latest.zip",
"hash": "ad595ba0a6e4ec2eb338e1724865c58892da847cb0f6ebbd3db22f3c10a9de33",
"pre_install": [
    "Rename-Item \"$dir\\Apifox-*\" 'Apifox.exe'",
    "Expand-7zipArchive \"$dir\\Apifox.exe\" \"$dir\" -Removal",
    "Expand-7zipArchive \"$dir\\`$PLUGINSDIR\\app-64.7z\" \"$dir\"",
    "Remove-Item \"$dir\\`$*\", \"$dir\\Uninstall*\" -Force -Recurse"
]
```

**Reply by Xelbayria** [reply]
Can you provide details on what are they doing? Starting from `"Rename-Item \"$dir\\Apifox-*\" 'Apifox.exe'"`

I understand that " * " is a wildcard, IIRC is what you call it.

**Reply by niheaven** [reply]
The key point is what is the archive format of the exe, and in the example there's version number that may be changed in different version, so the `Rename-Item`

Since the installer is an electronic one, two `Expand-7zipArchive` are used and the final `Remove-Item` cleans up unnecessary files.

**Comment by lewis-yeung** [comment]
You can write something like this:

```json
"url": "https://www.inputdirector.com/downloads/InputDirector.v2.1.2.zip",
"pre_install": "Rename-Item \"$dir\\InputDirector.*.Setup.exe\" \"$dir\\InputDirector.Setup.exe\"",
"installer": {
    "file": "InputDirector.Setup.exe",
    "args": "..."
}
```

See the manifest of [Extras/winaero-tweaker](https://github.com/ScoopInstaller/Extras/blob/master/bucket/winaero-tweaker.json) for example.

**Reply by Xelbayria** [reply]
Alright, thanks for providing some details and I'll use them for reference to see if i can get it working. A question, isn't it possible to use "innosetup" to extract from the exe that come from ZIP folder?

**Reply by lewis-yeung** [reply]
> A question, isn't it possible to use "innosetup" to extract from the exe that come from ZIP folder?

You can use the function [`Expand-InnoArchive`](https://github.com/ScoopInstaller/Scoop/blob/80b52e32a11fda2089949aaec00295d6185da3f7/lib/decompress.ps1#L185) in an installer script. See the manifest of [Versions/openssl-light](https://github.com/ScoopInstaller/Versions/blob/d7aa2dee125065eb77fa873f1171dc10ca7731d6/bucket/openssl-light.json#L17-L23) for example.

**Reply by rashil2000** [reply]
You'll have to use the `Expand-InnoArchive` command in pre_install

**Reply by Xelbayria** [reply]
alright. Are there more details on this part: (like what is "/SP-"? )
```JSON
"installer": {
        "file": "setup.exe",
        "args": [
            "/SP-",
            "/VERYSILENT",
            "/PORTABLE",
            "/DIR=\"$dir\""
        ]
    },
```
and do you have one more example of using Expand-InnoArchive? <strike> Also this `"Expand-InnoArchive -Path \"$dir\\$fname\""` what it does is open the EXE file like it's ZIP folder?  $dir is the current folder that EXE file is in? correct? </strike>

EDIT: 
Is there a good example where it extracts EXE from the ZIP folder and then use `Expand-InnoArchive` so it will install the files form EXE into $dir? 

EDIT 2: 
Just curious, is it possible to open ZIP folder and then rename EXE to 7z, where I can extract files from (EXE/7z) into $dir. I have yet to succeed because the documentation is too limited on how to use "installer".

**Reply by Xelbayria** [reply]
Ah yes! I got it working! 
this is the code I used: 
```JSON
	"pre_install": [
		"Get-ChildItem \"$dir\" '*.Setup.exe' | Rename-Item -NewName 'InputDirector.Setup.exe'",
		"Expand-7zipArchive \"$dir\\InputDirector.Setup.exe\" \"$dir\" -Removal"
	],
```
So I would like to learn more about the references and examples that you provided (above). I appreciate you teaching me more

**Reply by lewis-yeung** [reply]
> Are there more details on this part: (like what is "/SP-"? )

@Xelbayria This is an argument for the installer, which is app-specific.

**Reply by Xelbayria** [reply]
Oh ok. I've tried to use "Installer" on the program. I have yet to understand how it works and make it work. The program installer has 3 simple steps
1) agreement
2) directory installation (normally, it's `C:\Program Files (86x)`)
3) final check to make sure setup is all good and the button "install" 

that's it.

**Reply by rashil2000** [reply]
> Ah yes! I got it working! this is the code I used:
> 
> ```json
> 	"pre_install": [
> 		"Get-ChildItem \"$dir\" '*.Setup.exe' | Rename-Item -NewName 'InputDirector.Setup.exe'",
> 		"Expand-7zipArchive \"$dir\\InputDirector.Setup.exe\" \"$dir\" -Removal"
> 	],
> ```
> 
> So I would like to learn more about the references and examples that you provided (above). I appreciate you teaching me more

Does this program work if you launch the InputDirector.exe binary after this extraction?

**Comment by Xelbayria** [comment]
What's the difference between `script` and `args` via "installer"? 

I found the args for the installer of said program. 
/S is for silent installation
/D is for directory installation and must not have quotes even with spaces. 

I've tried to do this: 
```JSON
	"installer": {
		"file": "InputDirector.Setup.exe",
		"args": [
			"/S",
			"/D=$dir"
		]
	},
```
it failed. The error message below: 
`ERROR Exception calling "Start" with "0" argument(s): "The requested operation requires elevation"`

EDIT: 
I forgot to mention this. 
it can be used with CMD which goes like this: 
```CMD
InputDirector.Set.exe /S /D=Installation Directory
```

**Reply by lewis-yeung** [reply]
> ERROR Exception calling "Start" with "0" argument(s): "The requested operation requires elevation"

Seems that it requires admin privileges (run as admin).

**Reply by Xelbayria** [reply]
hmm Will `args` work the same way as the command line is used in CMD to install inputdirector? 
I wasn't sure whether `script` or `args` is ideal for the installer. I read the documentation on both.  as for `script`, wasn't it like this: 
```JSON
"installer": {
    "script": "InputDirector.Set.exe /S /D=Installation Directory"
}
```
as far as I understand, `args` can be only passed to the installer's options (enabling them or not) just like the normal activities you would do with the installer. 

EDIT: 
forgot to say that I tried to run Powershell as Admin, and it still gives me the same error. and The error message, I supposed that nothing of `args` was passed to the installer

**Reply by lewis-yeung** [reply]
If you use `script`, it can be something like this:

```json
"installer": {
    "script": "Invoke-ExternalCommand \"$dir\\InputDirector.Setup.exe\" -Args @('/S', \"/D=Installation Directory\") | Out-Null"
}
```

See the [docs](https://github.com/ScoopInstaller/Scoop/wiki/Pre-Post-(un)install-scripts) for variables that you can use in an installer script.

> forgot to say that I tried to run Powershell as Admin, and it still gives me the same error. and The error message, I supposed that nothing of `args` was passed to the installer

I don't know why it happened. Are you sure that all arguments for the installer are correct and you have a `pre_install` script to rename the installer executable to `InputDirector.Setup.exe`? Also, could you post a complete app manifest you wrote?

**Reply by Xelbayria** [reply]
Yes I'm 110% sure that the arguemnts for installer are correct. There is a FAQ on Input Director on how to install it via CMD which is this: 
`InputDirector.Set.exe /S /D=Installation Directory`  

installation Directory could be C:\Anywhere\Apps and it will create a folder named "input director" and extract the information into the folder. that's it. 
/S is a Silent mode. No need to mention what "/D" is... /D must be the last.  that means `/D=Directory /S` won't work. 

yes it has `pre_installer` but `installer` is yet to be determined and here's the code below: 
```JSON
{
    "version": "2.1.2",
    "description": "Input Director enables the control of multiple Windows systems using the keyboard/mouse attached to one computer",
    "homepage": "www.inputdirector.com",
    "license": "Freeware",
    "url": "https://www.inputdirector.com/downloads/InputDirector.v2.1.2.zip",
    "hash": "487052894233812c33c44bea48d92cfc123f88fb2a27e113a636c8219695bcc3",
    "pre_install": [
		"Get-ChildItem \"$dir\" '*.Setup.exe' | Rename-Item -NewName 'InputDirector.Setup.exe'"
	],
    "installer": {
		
	},
    "shortcuts": [
        [
            "InputDirector.exe",
	    "InputDirector"
        ]
    ],
}
```

**Reply by lewis-yeung** [reply]
I noticed that you mentioned two key points:

> /D is for directory installation and must not have quotes even with spaces.

> /D must be the last.

So I think the installer script below may not work as expected in a PowerShell context if the installation directory includes spaces.

> ```json
> "installer": {
>     "script": "Invoke-ExternalCommand \"$dir\\InputDirector.Setup.exe\" -Args @('/S', \"/D=Installation Directory\") | Out-Null"
> }
> ```

I guess the `InputDirector` installer executable has its own rules of parsing command line arguments from the entire raw command line string (something similar to what CMD does) instead of following the general MSVC Runtime's rules (see [MS docs](https://docs.microsoft.com/en-us/previous-versions/17w5ykft(v=vs.85)) for details).

Even worse, I think we cannot make it work using `args` due to this BAD DESIGN of how the installer executable parses command line arguments. **Either `args` or `installer` script, it will make a command line string from the given arguments, with any that includes spaces properly escaped by adding outer quotes, which breaks the rule of the unconventional command line parsing.**

**Reply by Xelbayria** [reply]
alas... So that's not good. 😔 
Looks like I'll have to speak to the author of Input Director about this. 
and how would you use `args` or `scripts` if it doesn't follow General MSVC Runtime Rule?  
I am going to find out if there is a way to make the `installer` use CMD to install inputdirector. 🤔

**Reply by lewis-yeung** [reply]
> and how would you use `args` or `scripts` if it doesn't follow General MSVC Runtime Rule?

I don't have any manifest on such case. No idea yet.

Changing ideas and to simplify the problem, the only and the simplest suggestion I can provide is to create a temporary batch (.bat) script using `Set-Content` cmdlet and `@"..."@` syntax in `installer.script` (see [this example](https://github.com/lewis-yeung/scoop-bucket/blob/6882f75b8af9008f905e8f8034c0244122435db4/bucket/msedgeredirect.json#L20-L34)) and remove it after the installation. If you don't like a temporary file, you should choose a way to extract files manually from `InputDirector.Setup.exe`, just as mentioned above.

> I am going to find out if there is a way to make the `installer` use CMD to install inputdirector.

No perfect solution for such case. Currently PowerShell cannot properly handle all cases of passing the entire command line string which contains CMD-defined special characters to CMD.

**Reply by Xelbayria** [reply]
> you should choose a way to extract files manually from `InputDirector.Setup.exe`, just as mentioned above.

I have tested extracting files manually from the installer, but it didn't work very well. It must be installed using the normal way (via the installer). You should have noticed that I managed to get it working by extracting files from the EXE. 

> No perfect solution for such case. Currently PowerShell cannot properly handle all cases of passing the entire command line string  which contains CMD-defined special characters to CMD.

That's why I'm speaking with the Author of the program to see if he can do something about the `/D` that should have quotes. perhaps that would sovle the isseu I am facing. I'm just waiting to hear from him. 

> the simplest suggestion I can provide is to create a temporary batch (.bat) script using Set-Content cmdlet and @"..."@ syntax in installer.script (see [this example](https://github.com/lewis-yeung/scoop-bucket/blob/6882f75b8af9008f905e8f8034c0244122435db4/bucket/msedgeredirect.json#L20-L34)) and remove it after the installation

Sorry, I do not fully understand what is above saying. 🤔  Are you saying that I could use `Set-Content cmdlet` to replace `Invoke-ExternalCommand`, correct?

**Reply by lewis-yeung** [reply]
> Are you saying that I could use `Set-Content cmdlet` to replace `Invoke-ExternalCommand`, correct?

No. I was saying that you can use `Set-Content` cmdlet to create a temporary batch script file, just as the [example](https://github.com/lewis-yeung/scoop-bucket/blob/6882f75b8af9008f905e8f8034c0244122435db4/bucket/msedgeredirect.json#L20-L34) I mentioned above. This way you can run the installer executable via CMD. The installer script can be like this:

```json
"installer": {
    "script": [
        "$install_dir = $dir -replace '%', '^%' # escape any '%' in the path",
        "Set-Content -Encoding utf8 -Path \"$dir\\install.bat\" -Value @\"",
        "@echo off",
        "InputDirector.Setup.exe /S /D=$install_dir",
        "\"@",
        "pushd \"$dir\"",
        "Invoke-ExternalCommand 'cmd.exe' -Args @('/c', \"$install_dir\\install.bat\") | Out-Null",
        "popd",
        "Remove-Item \"$dir\\install.bat\""
    ]
}
```

**Reply by Xelbayria** [reply]
> No. I was saying that you can use `Set-Content` cmdlet to create a temporary batch script file, just as the [example](https://github.com/lewis-yeung/scoop-bucket/blob/6882f75b8af9008f905e8f8034c0244122435db4/bucket/msedgeredirect.json#L20-L34) I mentioned above. This way you can run the installer executable via CMD. The installer script can be like this:
> 

Ok, I understand now. I've run some tests using the script above. it almost works... but the problem is that `Invoke-ExternalCommand 'cmd.exe` is not being called. I click on install.bat after removing `Remove-Item \"$dir\\install.bat\"` and got a "User Account Control" popup for my permission 
🤔 

EDIT: I almost forgot to mention the version of Powershell is v5.1, not 7.2.

**Reply by lewis-yeung** [reply]
> I click on install.bat after removing `Remove-Item \"$dir\\install.bat\"` and got a "User Account Control" popup for my permission

Try to change the `Invoke-ExternalCommand` line to

```
"Invoke-ExternalCommand 'cmd.exe' -Args @('/c', \"$install_dir\\install.bat\") -RunAs | Out-Null"
```

With `-RunAs` it will be run as admin. Since it requires elevation, a UAC popup is inevitable unless you have disabled it.

BTW, I believe that currently the function `Invoke-ExternalCommand` has a problem on concatenating arguments (but nothing to do with a UAC popup). I'll put up a PR to fix it.

**Reply by Xelbayria** [reply]
I ran a simple experiment where I have `hello.bat`, inside the file: 
```
@echo Hello World
pause
```
Simple enough, then i tried to run it via powershell with this command: `cmd.exe /C hello.bat`
and it doesn't open CMD.exe but show the result that CMD.exe would provide inside the PowerShell 🤔 

---------

anyway, I've done a quick test on the new `Invoke-ExternalCommand`, it works alright but the strange thing is that the CMD.exe is not running the command from "install.bat". I supposed that's what you meant by this: 

>BTW, I believe that currently the function Invoke-ExternalCommand has a problem with concatenating arguments (but nothing to do with a UAC popup). I'll put up a PR to fix it.

I understand what is "Concatenation" in coding.  No worries.  

EDIT:
Ah, of course, I see why. I understand what's the problem.

**Reply by lewis-yeung** [reply]
I just figured out the problem and finally got it work. I added `$install_dir\\` in front of `InputDirector.Setup.exe` in the temporary `install.bat` script. The complete installer script I used for the test:

```json
"installer": {
    "script": [
        "$install_dir = $dir -replace '%', '^%' # escape any '%' in the path",
        "Set-Content -Encoding utf8 -Path \"$dir\\install.bat\" -Value @\"",
        "@echo off",
        "$install_dir\\InputDirector.Setup.exe /S /D=$install_dir",
        "\"@",
        "Invoke-ExternalCommand 'cmd.exe' -Args @('/c', \"`\"$install_dir\\install.bat`\"\") -RunAs | Out-Null",
        "Remove-Item \"$dir\\install.bat\""
    ]
}
```

The reason is simple: no working directory will be set for the `cmd.exe` process when we use `-RunAs`, so CMD cannot find the executable `InputDirector.Setup.exe`... 😅

Besides, we must ensure that all arguments for `cmd.exe` have been manually escaped, so I added outer quotes for the temporary script path in case `$install_dir` contains any spaces.

**Reply by Xelbayria** [reply]
>The reason is simple: no working directory will be set for the cmd.exe process when we use -RunAs, so CMD cannot find the executable InputDirector.Setup.exe... 😅
Besides, we must ensure that all arguments for cmd.exe have been manually escaped, so I added outer quotes for the temporary script path in case $install_dir contains any spaces.

Yeah, I noticed the problem is that it's not finding the InputDirector.Setup.exe. Dang. I didn't think of adding the solution `$install_dir\\InputDirector.Setup.exe`. oh well. I didn't look at the code deep enough. 

Strange, it's not working on my side. My Powershell version is 5.1. I wasn't sure how to upgrade it to 6.1 or 7.2 🤔

**Reply by lewis-yeung** [reply]
> Strange, it's not working on my side. My Powershell version is 5.1.

I noticed that in PowerShell v5.1, `Set-Content -Encoding utf8 ...` output a file using "UTF-8 with BOM" encoding which messed up the command inside. So I tweaked the installer script:

```json
"installer": {
    "script": [
        "$install_dir = $dir -replace '%', '^%' # escape any '%' in the path",
        "Set-Content -Encoding Ascii -Path \"$dir\\install.bat\" -Value @\"",
        "@`\"$install_dir\\InputDirector.Setup.exe`\" /S /D=$install_dir",
        "\"@",
        "Invoke-ExternalCommand 'cmd.exe' -Args @('/c', \"`\"$install_dir\\install.bat`\"\") -RunAs | Out-Null",
        "Remove-Item \"$dir\\install.bat\""
    ]
}
```

Then I had a test with PowerShell v5.1 on Win11 and it worked fine on my end. The key is changing the encoding `utf8` to `Ascii`. It seems the annoying encoding problem doesn't appear in PowerShell v7.2.5. BTW, outer quotes for `$install_dir\\InputDirector.Setup.exe` are also indispensable.

> I wasn't sure how to upgrade it to 6.1 or 7.2

See [MS docs](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.2). Alternatively, you can run `scoop install main/pwsh` to install via Scoop.

**Reply by Xelbayria** [reply]
It almost worked... 😓 but the last part is not working. I removed `Remove-Item` for install.bat so I took a look inside the `install.bat` and here's what it say: 
```BAT
@"C:\Program Scoop\apps\inputdirector\2.1.2\InputDirector.Setup.exe" /S /D=C:\Program Scoop\apps\inputdirector\2.1.2
```
That can't be right 🤔 I also add "@echo Running Installer", just before the command above, so I can see what CMD is doing.. and it's working but I don't think it can find the `InputDirectory.Setup.exe`

just a reminder: My OS is Windows 10 Pro, not 11. I also tried both 5.1 and 7.2. It provided the same result. 

>See [MS docs](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.2). Alternatively, you can run scoop install main/pwsh to install via Scoop.

👍  that make it easier for me to get Powershell 7.2

EDIT: 
nvm. it turned out i need to uninstall the other Input Director on my computer cleanly before I can use Scoop to install it and now it's working. All is good now.

Just a question, is there a way to kill `IDWinService.exe` without having to go into Task Manager to do it. When Update/Cleanup happened, there will be a message saying that it cannot remove the folder because it's in use. ID Service was running in the background. it needs to be killed before updating or cleanup. The folder of Inputdirector has `uninstall.exe`, I need to check the documentation to see if there's a way to use `uninstall.exe`.

Also, there is another function of this `Inputdirector.Setup.exe /S` is that it also can silently upgrade the current of the folder. 🤔 
Is there a way I don't need to have a "version" folder, just using the "current" folder to run input-director? that would be good enough for using Inputdirector's installer

**Reply by lewis-yeung** [reply]
> Just a question, is there a way to kill `IDWinService.exe` without having to go into Task Manager to do it.

The [`Stop-Process`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/stop-process?view=powershell-7.2) cmdlet can be what you are looking for. Of course, I think the cleanup stuff should be left to the uninstaller (`uninstall.exe`) itself, if possible. This way we don't have to handle all cases ourselves, which simplifies the problem.

> Is there a way I don't need to have a "version" folder, just using the "current" folder to run input-director? that would be good enough for using Inputdirector's installer

EDITED: I don't quite understand the problem yet...

**Reply by Xelbayria** [reply]
>Is there a way I don't need to have a "version" folder, just using the "current" folder to run input-director? that would be good enough for using Inputdirector's installer

Legend: InputDirector = ID

The problem is that uninstall.exe will just remove the registered program from Windows 10 system and also required a reboot to completely remove all of the files.  That means `uninstall.exe` is not very useful when it comes to the "version" folder...  it's better to use the "current" folder. so we don't need to use `scoop cleanup ID` to remove the "old version" folder. When I ran installer.exe for ID in the past, what it does is check if the ID is already installed on my system, then it will resort to "upgrade" instead of "install", it's simply replacing/overwriting the old files in the ID folder. 

Above is the reason why I had to cleanly uninstall ID from my computer when I tried to use Scoop is that it doesn't install ID into  `Scoop/Apps/ID`. That means it's not possible for ID's installer to install into a newer "version" folder. 

EDIT: 
>The [Stop-Process](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/stop-process?view=powershell-7.2) cmdlet can be what you are looking for.

Thank you, You just give me an idea... I looked up the way to kill the process via CMD and it's possible. 
`taskkill /IM \"InputDirector.exe\" /IM \"IDWinService.exe\" /F`

**Reply by lewis-yeung** [reply]
> That means it's not possible for ID's installer to install into a newer "version" folder.

Are you talking about how to make it update this app (that means an older version already exists) correctly via Scoop? The `current` is just a junction which points to a specific version directory. It is linked after the installation/uninstallation and unlinked before upgrading the app. Anyway, all actual processes are inseparable from a version directory. The only idea I can come up with is checking the variable `$cmd` (see [docs](https://github.com/ScoopInstaller/Scoop/wiki/Pre-Post-(un)install-scripts)) in the `pre_install` script to see if we are in a `scoop update` context and if so, use the uninstaller to clean up the existing old version before installing a new one. However, I also noticed that it requires a reboot to finish the uninstallation, so I'm not sure if the next installation will be successful without a reboot.

**Reply by Xelbayria** [reply]
> Are you talking about how to make it update this app (that means an older version already exists) correctly via Scoop? The `current` is just a junction which points to a specific version directory. It is linked after the installation/uninstallation and unlinked before upgrading the app. Anyway, all actual processes are inseparable from a version directory. 

Yes, that's what I have in mind. Given that the installer will only check to see if ID is already installed or not. If it is, then it will just "upgrade" where it will overwrite these existing files. 

>The only idea I can come up with is checking the variable `$cmd` (see [docs](https://github.com/ScoopInstaller/Scoop/wiki/Pre-Post-(un)install-scripts)) in the `pre_install` script to see if we are in a `scoop update` context and if so, use the uninstaller to clean up the existing old version before installing a new one. However, I also noticed that it requires a reboot to finish the uninstallation, so I'm not sure if the next installation will be successful without a reboot.

It's the reason why I would like to use the first part of what you said above. I am just checking to see if there's a way to do the updating part on the existing "version" folder as the current folder... I am aware of the system between the "version" and "current" folder.

**Reply by lewis-yeung** [reply]
> I am just checking to see if there's a way to do the updating part on the existing "version" folder as the current folder...

Theoretically impossible, I think. 😅 The `current` will definitely be unlinked before an upgrade and relinked to the new version when done.

**Reply by Xelbayria** [reply]
> Theoretically impossible, I think. 😅 The `current` will definitely be unlinked before an upgrade and relinked to the new version when done.

Dang. I was hoping for that. Alright, thank you for teaching me a lot of things.

**Comment by niheaven** [comment]
Another suggestion that not related to the topic but the software you may use, for system tools such as this "InputDirector", installing it manually is much better 'cos scoop could not or hardly handle it perfectly.

**Comment by Xelbayria** [comment, marked as answer]
Forgot to close the discussion: 

the solution is: 
```JSON
"pre_install": [
	"Get-ChildItem \"$dir\" '*.Setup.exe' | Rename-Item -NewName 'InputDirector.Setup.exe'"
],
"installer": {
	"script": [
		"$install_dir = $dir -replace '%', '^%' # escape any '%' in the path",
		"Set-Content -Encoding Ascii -Path \"$dir\\install.bat\" -Value @\"",
		"taskkill /IM \"InputDirector.exe\" /IM \"IDWinService.exe\" /F",
		"@echo Running Installer",
		"@`\"$install_dir\\InputDirector.Setup.exe`\" /S /D=$install_dir",
		"\"@",
		"Invoke-ExternalCommand 'cmd.exe' -Args @('/C', \"`\"$install_dir\\install.bat`\"\") -RunAs | Out-Null",
		"Remove-Item \"$dir\\install.bat\""
	]
},
"uninstaller": {
	"file": "uninstall.exe"
},
```

It's to install inputdirector while put it into the directory of scoop. There is also the option where you could install inputdirector to its default directory while the scoop will be checking for any update. 

Thanks for helping out.
